dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+   * @returns {Promise<WriteResult>} A promise that resolves with the result of
+   * the write. Throws an error if the write fails.
+   *
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
+   *
+   * bulkWriter
+   *  .update(documentRef, {foo: 'bar'})
+   *  .then(result => {
+   *    console.log('Successfully executed write at: ', result);
+   *  })
+   *  .catch(err => {
+   *    console.log('Write failed with: ', err);
+   *  });
+   * });
+   */
+  update(
+    documentRef: DocumentReference,
+    dataOrField: UpdateData | string | FieldPath,
+    ...preconditionOrValues: Array<
+      {lastUpdateTime?: Timestamp} | unknown | string | FieldPath
+    >
+  ): Promise<WriteResult> {
+    const bulkCommitBatch = this.getEligibleBatch(documentRef);
+    bulkCommitBatch.refsInBatch.add(documentRef);
+    bulkCommitBatch.writeBatch.update(
+      documentRef,
+      dataOrField,
+      preconditionOrValues
+    );
+
+    return this.getResult(
+      bulkCommitBatch.tag,
+      bulkCommitBatch.writeBatch.opCount
+    );
+  }
+
+  /**
+   * Returns a Promise that resolves when there are no more pending writes. It
+   * never fails.
+   *
+   * The promise resolves immediately if there no pending writes. Otherwise, the
+   * Promise waits for all previously issued writes, but it does not wait for
+   * writes that were added after the method is called. If you want to wait for
+   * additional writes, call `waitForPendingWrites()` again.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await waitForPendingWrites().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  async waitForPendingWrites(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    this.sendBatches();
+    const writePromises = trackedBatches.map(batch => {
+      batch.pending = true;
+      return this.resultsMap.get(batch.tag)!.promise;
+    });
+    await Promise.all(writePromises);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = true;
+    for (const bulkBatch of this.batchQueue) {
+      if (bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+
+      if (
+        bulkBatch.pending === false &&
+        bulkBatch.writeBatch.opCount < BULK_WRITER_MAX_BATCH_SIZE &&
+        !bulkBatch.refsInBatch.has(ref)
+      ) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        logger(
```

Comment:
We could consider logging this to console.warn, since `logger` is off by default.